File: app/src/main/java/com/yunfie/illustia/settings/db/SettingsDao.java
package com.yunfie.illustia.settings.db;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import java.util.List;

@Dao
public interface SettingsDao {
    @Query("SELECT * FROM search_history ORDER BY position ASC")
    List<SearchHistoryEntity> getSearchHistory();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertSearchHistory(List<SearchHistoryEntity> items);

    @Query("DELETE FROM search_history")
    void clearSearchHistory();

    @Query("SELECT * FROM favorite_tags ORDER BY position ASC")
    List<FavoriteTagEntity> getFavoriteTags();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertFavoriteTags(List<FavoriteTagEntity> items);

    @Query("DELETE FROM favorite_tags")
    void clearFavoriteTags();

    @Query("SELECT * FROM view_history ORDER BY position ASC")
    List<ViewHistoryEntity> getViewHistory();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertViewHistory(List<ViewHistoryEntity> items);

    @Query("DELETE FROM view_history")
    void clearViewHistory();

    @Query("SELECT * FROM accounts ORDER BY position ASC")
    List<AccountEntity> getAccounts();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAccounts(List<AccountEntity> items);

    @Query("DELETE FROM accounts")
    void clearAccounts();
}

File: app/src/main/java/com/yunfie/illustia/settings/db/SettingsDao.kt
package com.yunfie.illustia.settings.db

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transaction

@Dao
@Suppress("TooManyFunctions")
interface SettingsDao {
    @Query("SELECT * FROM search_history ORDER BY position ASC")
    fun getSearchHistory(): List<SearchHistoryEntity>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertSearchHistory(items: List<SearchHistoryEntity>)

    @Query("DELETE FROM search_history")
    fun clearSearchHistory()

    @Query("SELECT * FROM favorite_tags ORDER BY position ASC")
    fun getFavoriteTags(): List<FavoriteTagEntity>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertFavoriteTags(items: List<FavoriteTagEntity>)

    @Query("DELETE FROM favorite_tags")
    fun clearFavoriteTags()

    @Query("SELECT * FROM view_history ORDER BY position ASC")
    fun getViewHistory(): List<ViewHistoryEntity>

    @Query("SELECT * FROM view_history ORDER BY position ASC LIMIT :limit")
    fun getViewHistory(limit: Int): List<ViewHistoryEntity>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertViewHistory(items: List<ViewHistoryEntity>)

    @Query("DELETE FROM view_history")
    fun clearViewHistory()

    @Query("SELECT * FROM accounts ORDER BY position ASC")
    fun getAccounts(): List<AccountEntity>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertAccounts(items: List<AccountEntity>)

    @Query("DELETE FROM accounts")
    fun clearAccounts()

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun upsertSavedIllust(item: SavedIllustEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun upsertSavedIllustPages(items: List<SavedIllustPageEntity>)

    @Query("SELECT * FROM saved_illusts ORDER BY savedAt DESC")
    fun getSavedIllusts(): List<SavedIllustEntity>

    @Transaction
    @Query("SELECT * FROM saved_illusts WHERE illustId = :illustId LIMIT 1")
    fun getSavedIllust(illustId: Long): SavedIllustWithPages?

    @Query("DELETE FROM saved_illust_pages WHERE illustId = :illustId")
    fun deleteSavedIllustPages(illustId: Long)

    @Query("DELETE FROM saved_illusts WHERE illustId = :illustId")
    fun deleteSavedIllust(illustId: Long)

    @Query("DELETE FROM saved_illust_pages")
    fun clearSavedIllustPages()

    @Query("DELETE FROM saved_illusts")
    fun clearSavedIllusts()
}
